doc/src/challenges/0012-nonzero.md

```diff
@@ -0,0 +1,100 @@
+# Challenge 12: Safety of `NonZero`
+
+- **Status:** Open
+- **Tracking Issue:** [Link to issue](https://github.com/model-checking/verify-rust-std/issues/71)
+- **Start date:** *2024-08-23*
+- **End date:** *2024-12-10*
+
+-------------------
+
+## Goal
+
+Verify the safety of `NonZero` in `core::num`.
+
+### Assumptions
+
+`new` and `get` leverage `transmute_unchecked`, so verifying the safety of these methods would require verifying that transmutations are safe. This task is out of scope for this challenge (instead, it's work for [Challenge 1](0001-core-transmutation.md)). For this challenge, for a transmutation from type `T` to type `U`, it suffices to write and verify a contract that `T` and `U` have the same size.
+
+You may assume that each `NonZeroInner` type upholds the safety conditions of the `ZeroablePrimitive` trait. Specifically, you need not verify that the integer primitives which implement `ZeroablePrimitive` are valid when 0, or that transmutations to the `Option` type are sound.
```

Comment:
I'm on the fence about this assumption, especially the latter part about `Option`. Would appreciate thoughts. The validity-when-0 part seems trivial (I'm not even sure how one would go about proving that a `u8`, for example, can in fact be zero). We could have them verify that for each primitive `T` implementing `ZeroablePrimitive`, the size of `NonZero<T>` does indeed equal the size of `Option<NonZero<T>>`.

Reply:
This assumption looks safe to me since it's guaranteed by the compiler.